plugins/volumeFilter.go

```diff
@@ -175,74 +175,95 @@ func (f *volumeFilter) Apply(ops []txnbuild.Operation, sellingOffers []hProtocol
-					opToReturn.Amount = fmt.Sprintf("%.7f", newAmountBeingSold)
-					keepSellingBase = true
-					newAmountString = ", newAmountString = " + opToReturn.Amount
-				}
-			}
-			log.Printf("volumeFilter:  selling (base units), price=%.8f amount=%.8f, keep = (projectedSoldInBaseUnits) %.7f <= %.7f (config.BaseAssetCapInBaseUnits): keepSellingBase = %v%s", sellPrice, amountValueUnitsBeingSold, projectedSoldInBaseUnits, *lp.baseAssetCapInBaseUnits, keepSellingBase, newAmountString)
-		} else {
-			keepSellingBase = true
-		}
+	// extracts from base or quote side, depending on filter
+	otb, tbb, cap, e := extractFilterInputs(dailyOTB, dailyTBBAccumulator, lp)
+	if e != nil {
+		return nil, fmt.Errorf("could not extract filter inputs from filter: %s", e)
+	}
 
-		if lp.baseAssetCapInQuoteUnits != nil {
-			projectedSoldInQuoteUnits := *dailyOTB.BaseAssetCapInQuoteUnits + *dailyTBBAccumulator.BaseAssetCapInQuoteUnits + (newAmountBeingSold * sellPrice)
-			keepSellingQuote = projectedSoldInQuoteUnits <= *lp.baseAssetCapInQuoteUnits
-			newAmountString := ""
-			if lp.mode == volumeFilterModeExact && !keepSellingQuote {
-				newAmount := (*lp.baseAssetCapInQuoteUnits - *dailyOTB.BaseAssetCapInQuoteUnits - *dailyTBBAccumulator.BaseAssetCapInQuoteUnits) / sellPrice
-				if newAmount > 0 {
-					newAmountBeingSold = newAmount
-					opToReturn.Amount = fmt.Sprintf("%.7f", newAmountBeingSold)
-					keepSellingQuote = true
-					newAmountString = ", newAmountString = " + opToReturn.Amount
-				}
-			}
-			log.Printf("volumeFilter: selling (quote units), price=%.8f amount=%.8f, keep = (projectedSoldInQuoteUnits) %.7f <= %.7f (config.BaseAssetCapInQuoteUnits): keepSellingQuote = %v%s", sellPrice, amountValueUnitsBeingSold, projectedSoldInQuoteUnits, *lp.baseAssetCapInQuoteUnits, keepSellingQuote, newAmountString)
-		} else {
-			keepSellingQuote = true
-		}
+	// if projected is under the cap, update the tbb and return the original op
+	projected := otb + tbb + offerAmount*capPrice
+	if projected <= cap {
+		*dailyTBBAccumulator.BaseAssetCapInBaseUnits += offerAmount
+		*dailyTBBAccumulator.BaseAssetCapInQuoteUnits += offerAmount * offerPrice
+		return op, nil
+	}
 
-		if keepSellingBase && keepSellingQuote {
-			// update the dailyTBB to include the additional amounts so they can be used in the calculation of the next operation
-			*dailyTBBAccumulator.BaseAssetCapInBaseUnits += newAmountBeingSold
-			*dailyTBBAccumulator.BaseAssetCapInQuoteUnits += (newAmountBeingSold * sellPrice)
-			return opToReturn, nil
-		}
-	} else {
-		// TODO buying side - we need to implement this to support buy side filters; extract common logic from the above sell side case
+	if lp.mode != volumeFilterModeExact {
+		return nil, nil
 	}
 
-	// we don't want to keep it so return the dropped command
-	return nil, nil
+	// if exact mode and with remaining capacity, update the op amount and return the op
+	// else, return nil
+	remainingCapacity := (cap - otb - tbb) / capPrice
```

Comment:
here too will need to tweak for buy side if it is different because of the complexities of how manage sell offer works. can add as a TODO if you want (like suggested above)